### PR TITLE
feat(sam3): Implement Phase 2 architecture (PromptEncoder + MaskDecoder)

### PR DESCRIPTION
**Pull request by arc-wonders**
## Summary

This PR implements **Phase 2 of the SAM-3 integration**, adding the remaining **architecture-only components** required for inference:

- Prompt encoder (points → sparse & dense embeddings)
- Mask decoder with cross-attention, multi-mask output, and IoU head

The implementation is **torch-only**, **inference-first**, and does not introduce a model wrapper or pretrained weights.

---

## Included

- New modules under `kornia/models/sam3/architecture/`
- Minimal, CPU-safe **shape-only tests** for prompt encoding and mask decoding
- Dynamic spatial shape inference (no hard-coded resolutions)

---

## Out of Scope

- Model wrapper (`SAM3`)
- Pretrained weights
- End-to-end inference pipeline
- Export / ONNX / compile support

---

## Validation

- All Phase-2 tests passing
- No Phase-1 regressions
- Full SAM-3 test suite passing

### CODE REVIEW

**Comment by arc-wonders** [comment]
Update on Copilot comments:

I’ve addressed the correctness and behavior issues flagged earlier (e.g. fixed spatially-rigid normalization and clarified/enforced single-mask behavior in Phase 2).

The remaining Copilot suggestions (gradcheck / BaseTester migration, expanded exception coverage, and full multi-mask hypernetwork logic) would significantly expand the scope beyond this Phase-2 PR, which is intentionally architecture-only and inference-first. To keep review cycles short and aligned with the phased SAM-3 plan, I’ve left those out for now.

**Comment by arc-wonders** [comment]
Hi @Edgar , just a gentle ping on this PR 
You’d mentioned earlier that you’d take a look when you get a chance. I wanted to flag that Kornia’s 15-day review policy would auto-close this in ~5–6 days, so I didn’t want it to slip through the cracks.

No rush if you’re busy , happy to address any comments or make changes whenever you get a chance. Thanks!

File: kornia/models/sam3/architecture/mask_decoder.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    Takes image embeddings and prompt embeddings to produce segmentation masks
+    and IoU predictions for the prompted regions.
+    """
+
+    def __init__(
+        self,
+        embed_dim: int = 256,
+        num_multimask_outputs: int = 3,
+        activation: str = "gelu",
+        iou_head_depth: int = 3,
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/models/sam3/architecture/mask_decoder.py:101]: what about this ?

File: kornia/models/sam3/architecture/mask_decoder.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+            embed_dim: Embedding dimension.
+            num_multimask_outputs: Number of mask outputs per prompt.
+            activation: Activation function name.
+            iou_head_depth: Depth of IoU prediction head.
+            iou_head_hidden_dim: Hidden dimension of IoU head.
+        """
+        super().__init__()
+        self.embed_dim = embed_dim
+        self.num_multimask_outputs = num_multimask_outputs
+        self.activation = activation
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/models/sam3/architecture/mask_decoder.py:116]: ?

File: kornia/models/sam3/architecture/mask_decoder.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        # NOTE: Phase 2 stub - mask_tokens and hypernetwork MLPs will be used in Phase 3 for multi-mask generation
+
+        # IoU prediction head
+        self.iou_prediction_head = nn.Sequential(
+            nn.Linear(embed_dim, iou_head_hidden_dim),
+            nn.ReLU(),
+            nn.Linear(iou_head_hidden_dim, iou_head_hidden_dim),
+            nn.ReLU(),
+            nn.Linear(iou_head_hidden_dim, num_multimask_outputs),
+        )
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/models/sam3/architecture/mask_decoder.py:138]: ?

File: tests/models/sam3/test_sam3_prompt_and_mask_architecture.py
Comment (edgarriba) [inline tests/models/sam3/test_sam3_prompt_and_mask_architecture.py:1]: filename

**Review by edgarriba — CHANGES_REQUESTED** [review]
still pending some copilot reviews

File: kornia/models/sam3/architecture/mask_decoder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arc-wonders) [reply kornia/models/sam3/architecture/mask_decoder.py:101]: In Phase-2 the IoU head depth is intentionally fixed; making it configurable would expand scope into architectural behavior decisions planned for a later phase.

File: kornia/models/sam3/architecture/mask_decoder.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arc-wonders) [reply kornia/models/sam3/architecture/mask_decoder.py:116]: This is intentionally kept as a forward-compatible API parameter in Phase-2.

Phase-2 focuses on architecture wiring and shape-correct inference; activation usage affects behavioral design choices, which are deferred to a later phase. I’ve kept the parameter to avoid breaking the interface when it’s wired in Phase-3.

**Comment by arc-wonders** [comment]
@edgarriba 
I’ve renamed the test file to remove the phase-specific naming, since that shouldn’t be part of the public OSS surface.

For the remaining Copilot comments (activation, iou_head_depth, and their usage), these are intentionally kept as forward-compatible API hooks. Phase-2 fixes the behavior to keep the implementation minimal and inference-focused; wiring these would expand scope into design decisions planned for a later phase.
Happy to adjust if you’d prefer these handled in this PR.
